tests/test.py

```diff
@@ -294,42 +303,123 @@ def run_terraform(*, image: str, final: bool = False):
+    # This test requires LEARNING_MODE to be true because the autodetect functionality traverses into the testing sub-directories, including those
+    # which are purposefully insecure, which otherwise would exit non-zero early, resulting in a limited set of logs.
     test_terraform_dir = tests_test_dir.joinpath("terraform")
-    # There is always one log for each security tool, regardless of if that
-    # tool is installed in the image being used.  If a tool is not in the PATH
-    # and executable, a log message indicating that is generated
-    LOG.debug("Testing autodetect mode")
+    # There is always one log for each security tool, regardless of if that tool is installed in the image being used.  If a tool is not in the PATH
+    # and executable, a log message indicating that is generated.
     number_of_security_tools = len(CONFIG["commands"]["terraform"]["security"])
-    number_of_testing_folders = len(
-        [dir for dir in test_terraform_dir.iterdir() if dir.is_dir()]
-    )
-    autodetect_environment = copy.deepcopy(informational_environment)
+    all_test_folders = [dir for dir in test_terraform_dir.rglob("*") if dir.is_dir()]
+    test_folders_containing_only_files = []
+    for directory in all_test_folders:
+        items_in_dir = directory.iterdir()
+        for item in items_in_dir:
+            if item.is_dir():
+                break
+        else:
+            test_folders_containing_only_files.append(directory)
+    number_of_testing_folders = len(test_folders_containing_only_files)
+    learning_mode_and_autodetect_environment = copy.deepcopy(informational_environment)
+    LOG.debug("Testing LEARNING_MODE with various AUTODETECT configurations")
     for autodetect_status in ["true", "false"]:
         if autodetect_status == "true":
             expected_number_of_logs = (
                 number_of_security_tools * number_of_testing_folders
             )
         else:
             expected_number_of_logs = number_of_security_tools
-        test_log_length = f"if [[ $(wc -l /var/log/easy_infra.log | awk '{{print $1}}') != {expected_number_of_logs} ]]; then exit 1; fi"
+        test_log_length = f"if [[ $(wc -l /var/log/easy_infra.log | awk '{{print $1}}') != {expected_number_of_logs} ]]; then exit 230; fi"
         command = f'/bin/bash -c "terraform init -backend=false && {test_log_length}"'
-        autodetect_environment["AUTODETECT"] = autodetect_status
+        learning_mode_and_autodetect_environment["AUTODETECT"] = autodetect_status
         utils.opinionated_docker_run(
             image=image,
             volumes=terraform_autodetect_volumes,
             command=command,
-            environment=autodetect_environment,
+            environment=learning_mode_and_autodetect_environment,
             expected_exit=0,
         )
         num_tests_ran += 1
 
+    # Ensure autodetect finds the appropriate terraform configs, which can be inferred by the number of logs written to /var/log/easy_infra.log
```

Comment:
This section exercises two angles of the change we just made; the exit codes and the appropriate amounts of logging.  The run command, check exit code is handled via the `opinionated_docker_run` helper function, but a docker run with `detach=True`, an `exec` command XYZ, and then measure the logs requires a bit more manual work.